Paltineanu Rares-Mihai/3.5 - curs5/main.c

```diff
@@ -0,0 +1,140 @@
+/**
+Write the functions for addition, subtraction and multiplication of two matrices,
+and then compute A=B*C�2*(B+C), where BandCare two quadratic matrices of order n.
+*/
+#include <stdio.h>
+#include <stdlib.h>
+/// 1 2 3   1 2 3
+/// 1 2 3   1 2 3
+/// 1 2 3   1 2 3
+int n;
+
+void readMat(int **B) {
+
+    for(int i = 0; i < n; i++) {
+        for(int j = 0; j < n; j++) {
+            scanf("%d", &B[i][j]);
+        }
+    }
+
+}
+
+void writeMat(int **B) {
+
+    for(int i = 0; i < n; i++) {
+        for(int j = 0; j < n; j++) {
+            printf("%d ", B[i][j]);
+        }
+        printf("\n");
+    }
+}
+
+void addition(int **B, int **C, int **newD) {
+
+    for(int i = 0; i < n; i ++) {
+        for(int j = 0; j < n; j++) {
+            newD[i][j] = B[i][j] + C[i][j];
+        }
+    }
+}
+
+void reinitialzingD(int **D) {
+
+    for(int i = 0; i < n; i++) {
+        memset(D[i], 0, n * 4);
+    }
+
+}
+
+void subtraction(int **B, int **C, int **newD) {
+
+    for(int i = 0; i < n; i ++) {
+        for(int j = 0; j < n; j++) {
+            newD[i][j] = B[i][j] - C[i][j];
+        }
+    }
+}
+
+void copyMat(int **cpyA, int ** B) {
+
+    for(int i = 0; i < n; i++) {
+        for(int j = 0; j < n; j++) {
+            cpyA[i][j] = B[i][j];
+        }
+    }
+}
+
+void multiplicationBy2(int **D) {
+
+    for(int i = 0; i < n; i++) {
+        for(int j = 0; j < n; j++) {
+            D[i][j] *= 2;
+        }
+    }
+}
+
+void multiplicationD(int **B, int **C, int **newD) {
+
+    int value = 0, rowsD = 0, colD = 0;
+
+    for(int i = rowsD; i < n; i++) {
+
+        for(int j = 0; j < n; j++) {
+
+            value = value + B[i][j]*C[j][colD];
+        }
+        newD[rowsD][colD] = value;
+        colD++;
+        value = 0;
+        if(colD < n) {
+            i--;
+        } else {
+            colD = 0;
+            rowsD++;
+        }
+
+    }
+
+}
+
+int main() {
```

Comment:
Great work modularizing everything and splitting it into functions!

I don't know if you want this, but, there's also a way to allocate the matrix in a function as well, as to not have repeated code. I don't have a compiler at hand but you should be able to write something like this:

```
void allocateMatrix(int *** matrix, int rows, int colums) {
    *matrix = (int **) malloc(sizeof(int*) * rows);
    for(int i = 0; i < rows; i++) {
        *(matrix[i]) = (int *) malloc(sizeof(int) * colums);
    }
}
```

and then you should be able to call that function from your main file:
```
...
int ** A, B, C, D;
allocateMatrix(&A, n, n);
allocateMatrix(&B, n, n);
allocateMatrix(&C, n, n);
allocateMatrix(&D, n, n);
```

Pretty sure it should work, but maybe I'm missing something - tell me if it does or doesn't!

Reply:
it's not working